syncstorage/storage/memcached.py

```diff
@@ -880,6 +931,85 @@ def del_item(self, userid, item):
             raise ItemNotFoundError
         return modified
 
+    def get_batches_key(self, userid):
+        return _key(userid, "c", self.collection, "batches")
+
+    def get_cached_batches(self, userid):
+        return self.cache.gets(self.get_batches_key(userid))
+
+    def create_batch(self, userid):
+        bdata, bcasid = self.get_cached_batches(userid)
+        batch = get_timestamp()
+        batchid = int(batch * 1000)
+        if not bdata:
+            bdata = {}
+        if batchid in bdata:
+            raise ConflictError
+        bdata[batchid] = {"user": userid,
+                          "modified": batch,
+                          # FIXME Rough guesstimate of the maximum
+                          #       reasonable life span of a batch
+                          "expires": int(batch) + 2 * 3600,
+                          "items": []}
+        key = self.get_batches_key(userid)
+        if not self.cache.cas(key, bdata, bcasid):
+            raise ConflictError
+        return batchid
+
+    def valid_batch(self, userid, batch):
+        ts = get_timestamp()
+        batchid = str(batch)
+        bdata, bcasid = self.get_cached_batches(userid)
+
+        if batchid not in bdata or bdata[batchid]["expires"] < ts:
+            self.close_batch(userid, batchid)
+            return False
+        return True
+
+    def append_items_to_batch(self, userid, batch, items):
+        modified = get_timestamp()
+        batchid = str(batch)
+        bdata, bcasid = self.get_cached_batches(userid)
+        # Invalid, closed, or expired batch
+        if (not bdata or
+                batchid not in bdata or
+                bdata[batchid]["expires"] <= int(modified)):
+            raise InvalidBatch(batch, modified, bdata)
+
+        bdata[batchid]["items"].extend(items)
```

Comment:
We should check that the total size of the batch has not exceeded our limit, otherwise clients will be allowed to just accumulate data in memcached indefinitely.  It might even be worth putting some limit on the number of active batches per userid as an additional safely measure.